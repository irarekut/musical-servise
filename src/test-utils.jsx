/* eslint-disable no-undef */
/* eslint-disable react/jsx-no-constructed-context-values */
import { useState } from 'react'
import { render, cleanup } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { setupListeners } from '@reduxjs/toolkit/query'
import { ThemeContext, themes } from './context/ContextTheme'
import { musicApi } from './api/api'
import authReducer from './store/slises/auth'
import filterReducer from './store/slises/filter'
import playerReducer from './store/slises/player'
import favoriteReducer from './store/slises/favorite'

function AllProviders({ children }) {
  const [currentTheme, setCurrentTheme] = useState(themes.light)

  const toggleTheme = () => {
    if (currentTheme === themes.dark) {
      setCurrentTheme(themes.light)
      return
    }

    setCurrentTheme(themes.dark)
  }

  return (
    <BrowserRouter>
      <ThemeContext.Provider value={{ theme: currentTheme, toggleTheme }}>
        {children}
      </ThemeContext.Provider>
    </BrowserRouter>
  )
}

export function withStoreProvider(store) {
  return function Wrapper({ children }) {
    return (
      <AllProviders>
        <Provider store={store}>{children}</Provider>
      </AllProviders>
    )
  }
}

export const setupApiStore = (api, withoutListeners) => {
  const getStore = () =>
    configureStore({
      reducer: {
        [musicApi.reducerPath]: musicApi.reducer,
        auth: authReducer,
        player: playerReducer,
        filter: filterReducer,
        favorite: favoriteReducer,
      },
      middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware().concat(musicApi.middleware),
    })
  const initialStore = getStore()
  const refObj = {
    api,
    store: initialStore,
    wrapper: withStoreProvider(initialStore),
  }

  let cleanupListeners

  beforeEach(() => {
    const store = getStore()
    refObj.store = store
    refObj.wrapper = withStoreProvider(store)

    if (!withoutListeners) {
      cleanupListeners = setupListeners(store.dispatch)
    }
  })

  afterEach(() => {
    cleanup()

    if (!withoutListeners) {
      cleanupListeners()
    }

    refObj.store.dispatch(api.util.resetApiState())
  })

  return refObj
}

export const customRender = (ui, options) =>
  render(ui, { wrapper: AllProviders, ...options })
