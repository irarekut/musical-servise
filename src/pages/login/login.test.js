/* eslint-disable no-restricted-globals */
/* eslint-disable no-undef */
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import { setupApiStore } from '../../test-utils'
import Login from './login'
import { musicApi } from '../../api/api'

const BASE_API_URL = 'https://painassasin.online/'

export const handlers = [
  rest.get(`${BASE_API_URL}login`, (req, res, ctx) =>
    res(ctx.json([{ token: '', isLogin: false }]))
  ),
]
const server = setupServer(...handlers)

// Мокируем api store
const storeRef = setupApiStore(musicApi)

describe('<Login />', () => {
  beforeAll(() => server.listen())

  // Чистим обработчики между тестами
  afterEach(() => server.resetHandlers())

  // Отрубаем сервер после выполнения тестов.
  // НЕ ЗАБЫВАЙТЕ ЭТО ДЕЛАТЬ, иначе сервер будет работать вхолостую
  afterAll(() => server.close())
  it('should render successfully and match snapshot', () => {
    const { container } = render(<Login />, { wrapper: storeRef.wrapper })

    expect(container).toMatchSnapshot()
  })

  it('should show requested data', async () => {
    render(<Login />, { wrapper: storeRef.wrapper })

    // Ждем ответа от сервера. Как только он придет,
    // отрисуется пункт списка
    expect(await screen.findByText('Войти')).toBeInTheDocument()
    expect(await screen.findByText('Зарегистрироваться')).toBeInTheDocument()
  })
})
