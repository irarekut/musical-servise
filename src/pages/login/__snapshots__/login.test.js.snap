// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should render successfully and match snapshot 1`] = `
<div>
  <form
    class="sc-jSUZER jBqxbm"
  >
    <div
      class="sc-hLBbgP"
    >
      <img
        alt="logo"
        class="sc-eDvSVe qkmhQ"
        src="logoLogin.png"
        style="width: 113.33px; height: 17px;"
      />
    </div>
    <div
      class="sc-gKPRtg ifzBUm"
    >
      <div
        class="sc-bcXHqe bRLOeJ"
      >
        <input
          class="sc-gswNZR fbvIkD"
          placeholder="e-mail"
          type="email"
          value=""
        />
        <span
          class="sc-dkrFOg PdUkF"
        />
      </div>
      <div
        class="sc-bcXHqe bRLOeJ"
      >
        <input
          class="sc-gswNZR fbvIkD"
          placeholder="Пароль"
          type="password"
          value=""
        />
        <span
          class="sc-dkrFOg PdUkF"
        />
      </div>
    </div>
    <button
      class="sc-jrcTuL bRcizN"
      type="submit"
    >
      Войти
    </button>
    <button
      class="sc-iBYQkv cgxitR"
    >
      Зарегистрироваться
    </button>
  </form>
</div>
`;
